### DISCUSSION TITLE
Dictionary not updating after MemberWiseClone

### DISCUSSION THREAD

**Discussion by MaxTotal** [dotnet/runtime#103075] [Q&A]
Hello,

I did something that we could resume like this today : 

```
public class myObject {

private Dictionary<string, string> parameters = new Dictionary<string, string>();

public myObject DeepCopy()
{
    myObject other = (myObject)this.MemberwiseClone();
    return other;
}
}

myObject api = new myObject();
api.Add("dd", "17");
myObject apiCopy = api.DeepCopy();
apiCopy.Parameters["dd"] = "4";
```

After this, the value of my Parameters["dd"] in apiCopy is still 17 whereas I did a MemberwiseClone(). It is very confusing that the value didn't change whereas the object was supposed to be cloned.

**Comment by HaloFour** [comment]
MemberwiseClone creates a [shallow copy](https://learn.microsoft.com/en-us/dotnet/api/system.object.memberwiseclone?view=net-8.0#remarks), not a deep copy.  The Dictionary itself is not cloned, only myObject is.

This is also a question for the [runtime repository](https://github.com/dotnet/runtime).

**Reply by MaxTotal** [reply]
So there is no easy way to do a deep copy for an object with Dictionary right ? I tried reflection and Serialization, both didn't give any satisfying result.

**Reply by HaloFour** [reply]
There is no built-in facility to perform a deep copy of an instance in .NET.  If you want to copy the contents of that dictionary into another dictionary, that would be on the `DeepCopy` method.  Serialization should result in that happening as deserialization will make and populate a new instance of a dictionary, although that would be a fairly expensive way to attempt to handle this.

**Reply by CyrusNajmabadi** [reply]
If you're using reflection, you can see if you have a dictionary, and then make a new dictionary from the original.  Dictionary has a cosntructor that will copy the keys/values from another collection to initialize itself.

**Reply by elgonzo** [reply]
Strictly speaking, with the given example it's not even necessary to use serialization/deserialization or reflection.

The `DeepCopy()` method being a member of the `myObject`<sup>[1]</sup> class allows it to access private members of _any_ `myObject` instances. And for dictionaries where you don't need to clone the values in the dictionary themselves (as should be the case with strings), the DeepCopy method could handle cloning of the dictionary directly pretty much like this:
```C#
public myObject DeepCopy()
{
    // With respect to deep cloning, MemberwiseClone is only useful here if myObject has value type fields that need to be copied.
    // If myObject has no value type fields, the call to MemberwiseClone could be replaced with new myObject(), as it doesn't help with deep cloning of reference type fields.
    myObject other = (myObject) this.MemberwiseClone();
    other.parameters = new(this.parameters);
    return other;
}
```

------

<sup>[1]</sup> "myObject" is a confusing name for something that is a class/type and not an object/instance. Yeah, i know, there is also the precedence of the System.Object type existing in the BCL, which imo makes reading this always funny: `new Object()`, i.e., "create an Object object".

**Comment by jnm2** [comment]
Here's a compiling version of that code snippet, demonstrating that "the value of my Parameters["dd"] in apiCopy" is 4, not 17, when this code runs: https://sharplab.io/#v2:C4LgTgrgdgPgAgJgIwFgBQcAMACOSAsA3OugLYCeA8gEYBWApgMbDYCGADgJbYC82U9AO7YKNBswAUASmJoOnAHQAFVmFal6wemADOCgIIATQxIBEx0wBpsppAHZTMslTpMW8gMIB7duV5suBQARenp2b19pWU8fcmVVdU1tHQBtc0NTAF1/U3xTWTwATgkY33i1DS1dNItMpzR0OABmXAQRF3EWAG8SDBagzmZOLyhVcgAePExrKYA+bBUKpN1/AWEBoZGxyaRp3F3ZqN7m9rE3bBCwiPJpdB60bEfT12ZsL2AAC21/CVEX4Ckn04egAsvRSNRtIJgfQPAAbEb0I4PJ5wOxvT7aWQAX3QuLQQA=

**Reply by KennethHoff** [reply]
For reference, `api` and `apiCopy` both are 4 in this scenario.
